File: src/patterns/structural/bridge/remote.nim
# internal
import
    patterns/structural/bridge/[
        device
    ]


type
    Remote*[TDevice: Device] = ref object of RootObj
        device: TDevice


proc newRemote*[TDevice: Device](device: TDevice): Remote[TDevice] =
    return Remote[TDevice](device: device)

proc tooglePower*(self: Remote): bool =
    return if self.device.isEnabled(): self.device.disable() else: self.device.enable()

proc volumeDown*(self: Remote): bool =
    return self.device.setVolume(self.device.getVolume() - 1)

proc volumeUp*(self: Remote): bool =
    return self.device.setVolume(self.device.getVolume() + 1)

proc channelDown*(self: Remote): bool =
    return self.device.setChannel(self.device.getChannel() - 1)

proc channelUp*(self: Remote): bool =
    return self.device.setChannel(self.device.getChannel() + 1)

proc getDevice*(self: Remote): Device =
    return self.device

proc setDevice*(self: Remote, device: Device) =
    self.device = device

File: src/patterns/structural/bridge/remote.go
package bridge


type Remote struct {
	device Device
}


func NewRemote(device Device) *Remote {
	return &Remote{device: device}
}

func (self *Remote) TooglePower() bool {
	if (!self.device.IsEnabled()) {
		return self.device.Enable()
	}
	return self.device.Enable()
}

func (self *Remote) VolumeDown() bool {
	return self.device.SetVolume(self.device.GetVolume() - 1)
}

func (self *Remote) VolumeUp() bool {
	return self.device.SetVolume(self.device.GetVolume() + 1)
}

func (self *Remote) ChannelDown() bool {
	return self.device.SetChannel(self.device.GetChannel() - 1)
}

func (self *Remote) ChannelUp() bool {
	return self.device.SetChannel(self.device.GetChannel() + 1)
}

func (self *Remote) GetDevice() Device {
	return self.device
}

func (self *Remote) SetDevice(device Device) {
	self.device = device
}